go/simple/v2/ampli/ampli.go

```diff
@@ -0,0 +1,389 @@
+	stronglyTypedEvent.OptionalString = optionalString
+	stronglyTypedEvent.optionals["OptionalString"] = true
+
+	return stronglyTypedEvent
+}
+
+func (stronglyTypedEvent *EventWithAllProperties) SetOptionalBool(optionalBool bool) *EventWithAllProperties {
+	stronglyTypedEvent.OptionalBool = optionalBool
+	stronglyTypedEvent.optionals["OptionalBool"] = true
+
+	return stronglyTypedEvent
+}
+
+func (stronglyTypedEvent EventWithAllProperties) ToEvent() amplitude.Event {
+	event := amplitude.Event{
+		EventType: "Event With All Properties",
+		EventProperties: map[string]interface{}{
+			"requiredArray":   stronglyTypedEvent.RequiredArray,
+			"requiredBool":    stronglyTypedEvent.RequiredBool,
+			"requiredEnum":    stronglyTypedEvent.RequiredEnum,
+			"requiredInteger": stronglyTypedEvent.RequiredInteger,
+			"requiredNumber":  stronglyTypedEvent.RequiredNumber,
+			"requiredString":  stronglyTypedEvent.RequiredString,
+		},
+	}
+
+	if _, present := stronglyTypedEvent.optionals["OptionalString"]; present {
+		event.EventProperties["optionalString"] = stronglyTypedEvent.OptionalString
+	}
+
+	if _, present := stronglyTypedEvent.optionals["OptionalBool"]; present {
+		event.EventProperties["optionalBool"] = stronglyTypedEvent.OptionalBool
+	}
+
+	return event
+}
+
+type Ampli struct {
+	Disabled bool
+	Client   amplitude.Client
+	mutex    sync.RWMutex
+}
+
+// Load initializes the Ampli wrapper.
+// Call once when your application starts.
+func (a *Ampli) Load(options LoadOptions) {
+	if a.Client != nil {
+		log.Default().Printf("Warn: Ampli is already initialized. Ampli.load() should be called once at application start up.")
+
+		return
+	}
+
+	if options.Client.isEmpty() {
+		options.Client = LoadClientOptions{
+			Configuration: DefaultConfiguration(),
+		}
+	}
+
+	var apiKey string
+	if options.Client.APIKey != "" {
+		apiKey = options.Client.APIKey
+	} else if options.Environment != "" {
+		apiKey = APIKey[options.Environment]
+	} else if options.Client.Configuration.APIKey != "" {
+		apiKey = options.Client.Configuration.APIKey
+	}
+
+	if !(apiKey != "" || options.Client.Instance != nil) {
+		log.Default().Printf("Error: Ampli.Load() requires option.Environment, " +
+			"and apiKey from either options.Instance.APIKey or APIKey[options.Environment], " +
+			"or options.Instance.Instance")
+	}
+
+	var configuration amplitude.Config
+	if !options.Client.Configuration.IsEmpty() {
+		configuration = options.Client.Configuration
+	} else {
+		configuration = DefaultConfiguration()
+	}
+
+	if configuration.Plan == (amplitude.Plan{}) {
+		configuration.Plan = DefaultConfiguration().Plan
+	}
+
+	if options.Client.Instance != nil {
+		a.Client = options.Client.Instance
+	} else {
+		configuration.APIKey = apiKey
+		a.Client = amplitude.NewClient(configuration)
+	}
+
+	a.mutex.Lock()
+	a.Disabled = options.Disabled
+	a.mutex.Unlock()
+}
+
+// InitializedAndEnabled checks if Ampli is initialized and enabled.
+func (a *Ampli) InitializedAndEnabled() bool {
```

Comment:
TypeScript version:
```
  private isInitializedAndEnabled(): boolean {
    if (!this.amplitude) {
      console.error('ERROR: Ampli is not yet initialized. Have you called ampli.load() on app start?');
      return false;
    }
    return !this.disabled;
  }
```

Reply:
It is better to distinguish these two situations and explicitly tell users that client is not initialized.

Updated